### PR TITLE
Prometheus 2.45.3 GMP support

### PR DESCRIPTION
**Pull request by TheSpiritXIII**
https://github.com/GoogleCloudPlatform/prometheus-engine/pull/772

### CODE REVIEW

File: .gcloudignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.git
```
Comment (pintohutch) [inline .gcloudignore:1]: nit (non-blocking): What's this file for?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -61,25 +61,25 @@ jobs:
...
-    runs-on: ubuntu-latest
-    # The go verson in this image should be N-1 wrt test_go.
-    container:
-      image: quay.io/prometheus/golang-builder:1.20-base
-    steps:
-      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
-      - run: make build
-      - run: go test ./tsdb/...
-      - run: go test ./tsdb/ -test.tsdb-isolation=false
+  # test_golang_oldest:
```
Comment (pintohutch) [inline .github/workflows/ci.yml]: question: I imagine something breaks with this?

Is there a reason why we need it commented out now, as opposed to before?

File: .gitignore [hunk 3]
```diff
@@ -24,10 +24,9 @@ benchmark.txt
 npm_licenses.tar.bz2
 /web/ui/static/react
 
-/vendor
 /.build
 
-/**/node_modules
+/**/node_modules/.cache
```
Comment (pintohutch) [inline .gitignore:29]: question: I see this was checked in with f3c5df45bbfb88120b2b249f61c75125e10ec4ae, why did we need to change the file at all?

Comment (pintohutch) [reply .gitignore:29]: suggestion: could you add a comment explaining?

File: go.mod [hunk 4]
```diff
@@ -202,6 +202,8 @@ require (
 )
 
 replace (
+	// This version is a hard-requirement for content negotiation.
+	github.com/prometheus/common => github.com/TheSpiritXIII/prometheus-common v0.45.0-gmp.0
```
Comment (pintohutch) [inline go.mod:209]: question: Is this change necessary for the 2.45 fork?

IIUC we're not supporting secrets management with this PR, correct?

**Review by pintohutch — COMMENTED** [review]
Mostly nits and questions - thanks for putting this together!

File: go.mod (hunk 4 above)
Comment (TheSpiritXIII) [reply go.mod:209]: I updated the comment to be clearer. Let me know if it doesn't make sense!

File: .gcloudignore (hunk 1 above)
Comment (TheSpiritXIII) [reply .gcloudignore:1]: To prevent uploading the entire Git repository to Google Cloud Build.  This line is standard in a .gcloudignore file because when you're building, you don't care about the git history. Our file matches the official example: https://cloud.google.com/sdk/gcloud/reference/topic/gcloudignore

I believe that prom-bench would be utilizing this. However, the original change predates our use of prom-bench: https://github.com/GoogleCloudPlatform/prometheus/commit/623141cf6e2b9b6aad2e1a52ad1f6cb7152c9e9c

File: .github/workflows/ci.yml (hunk 2 above)
Comment (TheSpiritXIII) [reply .github/workflows/ci.yml]: Prometheus' CI builds with two Go versions: the latest and the one specified in the `go.mod`. At the time of creating this PR, prometheus-engine was using 1.21, which meant this fork needed to use the same version. However, since 1.21 was the latest and the oldest we could use, I commented this line out.

Now that 1.22 is available, I've updated this repository. Please take another look. Thanks!

File: .gitignore (hunk 3 above)
Comment (TheSpiritXIII) [reply .gitignore:29]: The same way `go mod vendor` downloads and stores all files in `vendor`, NodeJS downloads and stores all files in `node_modules`. Since we want to contain all vendoring within this repository, we allow `node_modules` as we do `vendor`.  We don't need any more than that to build, so we exclude the `.cache` directory. Otherwise, this directory would grow as we continue doing NodeJS and dependency upgrades.

Does the commit text not already explain the intent?

This change was originally part of the first-ever fork commit: https://github.com/GoogleCloudPlatform/prometheus/commit/623141cf6e2b9b6aad2e1a52ad1f6cb7152c9e9c

File: .gcloudignore (hunk 1 above)
Comment (pintohutch) [reply .gcloudignore:1]: Oh interesting. Ok - the git blame was confusing here. I'm assuming we technically don't need that anymore. It was probably added for doing manual cloud builds. But happy to keep for now, given precedent.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (pintohutch) [reply .github/workflows/ci.yml]: SGTM

File: .gitignore (hunk 3 above)
Comment (pintohutch) [reply .gitignore:29]: Ah makes sense. Thanks for clarifying!

File: go.mod (hunk 4 above)
Comment (pintohutch) [reply go.mod:209]: Discussed offline. tl;dr - this is already folded in our 2.43 fork and enables unit tests to pass.

**Review by pintohutch — APPROVED** [review]
LGTM!

Can we do a `/prombench` run?

**Comment by TheSpiritXIII** [comment]
Seeing about at most a 15% reduction in memory with high samples exported. No noticeable CPU usage change.

**Comment by bwplotka** [comment]
I see 10% on collector setup with our export on. Still, great progress!

image
